### PR TITLE
IGNITE-22101 Attempt to fix performance drop on async thin client requests

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Review by petrov-mg — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java [hunk 1]
```diff
@@ -118,9 +121,20 @@ private ClientListenerResponse handle0(ClientListenerRequest req) {
         if (req0.isAsync(ctx)) {
             IgniteInternalFuture<ClientResponse> fut = req0.processAsync(ctx);
 
+            if (!fut.isDone()) {
+                try {
+                    // Give request a chance to be executed and response processed by the current thread,
+                    // so we can avoid any performance drops caused by async requests execution.
+                    fut.get(ASYNC_REQUEST_WAIT_TIMEOUT);
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java:128]: Is it a good idea to block the IO thread, even for a short period? 

Do we have any benchmark results for this change?

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java:128]: @ptupitsyn, it's not IO thread, it's client-connector thread. Before IGNITE-21183 these threads are always blocked, and it seems to be ok except the cases with waiting on locked keys. Now we wait extra time only on locked keys, in other cases we release the thread as soon as result become available. 
Yes, I have benchmark results for this change, performance drop is fixed now. I will publish results today as a comment to the ticket.

Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java:128]: > not IO thread, it's client-connector thread

`client-connector` does IO, that's what I had in mind.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java [hunk 2]
```diff
@@ -45,6 +45,9 @@
  * Thin client request handler.
  */
 public class ClientRequestHandler implements ClientListenerRequestHandler {
+    /** Timeout to wait for async requests completion, to handle them as regular sync requests. */
+    private static final long ASYNC_REQUEST_WAIT_TIMEOUT = 10L;
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequestHandler.java:49]:
```suggestion
    private static final long ASYNC_REQUEST_WAIT_TIMEOUT_MILLIS = 10L;
```

**Review by ptupitsyn — APPROVED** [review]

**Pull request closed without merging**